README
NAME
    Web::Simple - A quick and easy way to build simple web applications

SYNOPSIS
      #!/usr/bin/env perl

      package HelloWorld;
      use Web::Simple;

      sub dispatch_request {
        sub (GET) {
          [ 200, [ 'Content-type', 'text/plain' ], [ 'Hello world!' ] ]
        },
        sub () {
          [ 405, [ 'Content-type', 'text/plain' ], [ 'Method not allowed' ] ]
        }
      }

      HelloWorld->run_if_script;

    If you save this file into your cgi-bin as "hello-world.cgi" and then
    visit:

      http://my.server.name/cgi-bin/hello-world.cgi/

    you'll get the "Hello world!" string output to your browser. At the same
    time this file will also act as a class module, so you can save it as
    HelloWorld.pm and use it as-is in test scripts or other deployment
    mechanisms.

    Note that you should retain the ->run_if_script even if your app is a
    module, since this additionally makes it valid as a .psgi file, which
    can be extremely useful during development.

    For more complex examples and non-CGI deployment, see
    Web::Simple::Deployment. To get help with Web::Simple, please connect to
    the irc.perl.org IRC network and join #web-simple.

DESCRIPTION
    The philosophy of Web::Simple is to keep to an absolute bare minimum for
    everything. It is not designed to be used for large scale applications;
    the Catalyst web framework already works very nicely for that and is a
    far more mature, well supported piece of software.

    However, if you have an application that only does a couple of things,
    and want to not have to think about complexities of deployment, then
    Web::Simple might be just the thing for you.

    The only public interface the Web::Simple module itself provides is an
    "import" based one:

      use Web::Simple 'NameOfApplication';

    This sets up your package (in this case "NameOfApplication" is your
    package) so that it inherits from Web::Simple::Application and imports
    strictures, as well as installs a "PSGI_ENV" constant for convenience,
    as well as some other subroutines.

    Importing strictures will automatically make your code use the "strict"
    and "warnings" pragma, so you can skip the usual:

      use strict;
      use warnings FATAL => 'aa';

    provided you 'use Web::Simple' at the top of the file. Note that we turn
    on *fatal* warnings so if you have any warnings at any point from the
    file that you did 'use Web::Simple' in, then your application will die.
    This is, so far, considered a feature.

    When we inherit from Web::Simple::Application we also use Moo, which is
    the the equivalent of:

      {
        package NameOfApplication;
        use Moo;
        extends 'Web::Simple::Application';
      }

    So you can use Moo features in your application, such as creating
    attributes using the "has" subroutine, etc. Please see the documentation
    for Moo for more information.

    It also exports the following subroutines for use in dispatchers:

      response_filter { ... };

      redispatch_to '/somewhere';

    Finally, import sets

      $INC{"NameOfApplication.pm"} = 'Set by "use Web::Simple;" invocation';

    so that perl will not attempt to load the application again even if

      require NameOfApplication;

    is encountered in other code.

DISPATCH STRATEGY
    Web::Simple despite being straightforward to use, has a powerful system
    for matching all sorts of incoming URLs to one or more subroutines.
    These subroutines can be simple actions to take for a given URL, or
    something more complicated, including entire Plack applications,
    Plack::Middleware and nested subdispatchers.

  Examples
     sub dispatch_request {
       # matches: GET /user/1.htm?show_details=1
       #          GET /user/1.htm
       sub (GET + /user/* + ?show_details~ + .htm|.html|.xhtml) {
         my ($self, $user_id, $show_details) = @_;
         ...
       },
       # matches: POST /user?username=frew
       #          POST /user?username=mst&first_name=matt&last_name=trout
       sub (POST + /user + ?username=&*) {
          my ($self, $username, $misc_params) = @_;
         ...
       },
       # matches: DELETE /user/1/friend/2
       sub (DELETE + /user/*/friend/*) {
         my ($self, $user_id, $friend_id) = @_;
         ...
       },
       # matches: PUT /user/1?first_name=Matt&last_name=Trout
       sub (PUT + /user/* + ?first_name~&last_name~) {
         my ($self, $user_id, $first_name, $last_name) = @_;
         ...
       },
       sub (/user/*/...) {
         my $user_id = $_[1];
         # matches: PUT /user/1/role/1
         sub (PUT + /role/*) {
           my $role_id = $_[1];
           ...
         },
         # matches: DELETE /user/1/role/1
         sub (DELETE + /role/*) {
           my $role_id = $_[1];
           ...
         },
       },
     }

  The dispatch cycle
    At the beginning of a request, your app's dispatch_request method is
    called with the PSGI $env as an argument. You can handle the request
    entirely in here and return a PSGI response arrayref if you want:

      sub dispatch_request {
        my ($self, $env) = @_;
        [ 404, [ 'Content-type' => 'text/plain' ], [ 'Amnesia == fail' ] ]
      }

    However, generally, instead of that, you return a set of dispatch subs:

      sub dispatch_request {
        my $self = shift;
        sub (/) { redispatch_to '/index.html' },
        sub (/user/*) { $self->show_user($_[1]) },
        ...
      }

    Well, a sub is a valid PSGI response too (for ultimate streaming and
    async cleverness). If you want to return a PSGI sub you have to wrap it
    into an array ref.

      sub dispatch_request {
        [ sub { 
            my $respond = shift;
            # This is pure PSGI here, so read perldoc PSGI
        } ]
      }

    If you return a subroutine with a prototype, the prototype is treated as
    a match specification - and if the test is passed, the body of the sub
    is called as a method any matched arguments (see below for more
    details).

    You can also return a plain subroutine which will be called with just
    $env - remember that in this case if you need $self you -must- close
    over it.

    If you return a normal object, Web::Simple will simply return it upwards
    on the assumption that a response_filter (or some arbitrary
    Plack::Middleware) somewhere will convert it to something useful. This
    allows:

      sub dispatch_request {
        my $self = shift;
        sub (.html) { response_filter { $self->render_zoom($_[0]) } },
        sub (/user/*) { $self->users->get($_[1]) },
      }

    to render a user object to HTML, if there is an incoming URL such as:

      http://myweb.org/user/111.html

    This works because as we descend down the dispachers, we first match
    "sub (.html)", which adds a "response_filter" (basically a specialized
    routine that follows the Plack::Middleware specification), and then
    later we also match "sub (/user/*)" which gets a user and returns that
    as the response. This user object 'bubbles up' through all the wrapping
    middleware until it hits the "response_filter" we defined, after which
    the return is converted to a true html response.

    However, two types of object are treated specially - a Plack::App object
    will have its "-"to_app> method called and be used as a dispatcher:

      sub dispatch_request {
        my $self = shift;
        sub (/static/...) { Plack::App::File->new(...) },
        ...
      }

    A Plack::Middleware object will be used as a filter for the rest of the
    dispatch being returned into:

      ## responds to /admin/track_usage AND /admin/delete_accounts

      sub dispatch_request {
        my $self = shift;
        sub (/admin/**) {
          Plack::Middleware::Session->new(%opts);
        },
        sub (/admin/track_usage) {
          ## something that needs a session
        },
        sub (/admin/delete_accounts) {
          ## something else that needs a session
        },
      }

    Note that this is for the dispatch being -returned- to, so if you want
    to provide it inline you need to do:

      ## ALSO responds to /admin/track_usage AND /admin/delete_accounts

      sub dispatch_request {
        my $self = shift;
        sub (/admin/...) {
          sub {
            Plack::Middleware::Session->new(%opts);
          },
          sub (/track_usage) {
            ## something that needs a session
          },
          sub (/delete_accounts) {
            ## something else that needs a session
          },
        }
      }

    And that's it - but remember that all this happens recursively - it's
    dispatchers all the way down. A URL incoming pattern will run all
    matching dispatchers and then hit all added filters or
    Plack::Middleware.

  Web::Simple match specifications
   Method matches
      sub (GET) {

    A match specification beginning with a capital letter matches HTTP
    requests with that request method.

   Path matches
      sub (/login) {

    A match specification beginning with a / is a path match. In the
    simplest case it matches a specific path. To match a path with a
    wildcard part, you can do:

      sub (/user/*) {
        $self->handle_user($_[1])

    This will match /user/<anything> where <anything> does not include a
    literal / character. The matched part becomes part of the match
    arguments. You can also match more than one part:

      sub (/user/*/*) {
        my ($self, $user_1, $user_2) = @_;

      sub (/domain/*/user/*) {
        my ($self, $domain, $user) = @_;

    and so on. To match an arbitrary number of parts, use -

      sub (/page/**) {
        my ($self, $match) = @_;

    This will result in a single element for the entire match. Note that you
    can do

      sub (/page/**/edit) {

    to match an arbitrary number of parts up to but not including some final
    part.

    Note: Since Web::Simple handles a concept of file extensions, * and **
    matchers will not by default match things after a final dot, and this
    can be modified by using *.* and **.* in the final position, i.e.:

      /one/*       matches /one/two.three    and captures "two"
      /one/*.*     matches /one/two.three    and captures "two.three"
      /**          matches /one/two.three    and captures "one/two"
      /**.*        matches /one/two.three    and captures "one/two.three"

    Finally,

      sub (/foo/...) {

    Will match "/foo/" on the beginning of the path -and- strip it. This is
    designed to be used to construct nested dispatch structures, but can
    also prove useful for having e.g. an optional language specification at
    the start of a path.

    Note that the '...' is a "maybe something here, maybe not" so the above
    specification will match like this:

      /foo         # no match
      /foo/        # match and strip path to '/'
      /foo/bar/baz # match and strip path to '/bar/baz'

    Almost the same,

      sub (/foo...) {

    Will match on "/foo/bar/baz", but also include "/foo". Otherwise it
    operates the same way as "/foo/...".

      /foo         # match and strip path to ''
      /foo/        # match and strip path to '/'
      /foo/bar/baz # match and strip path to '/bar/baz'

    Please note the difference between "sub(/foo/...)" and "sub(/foo...)".
    In the first case, this is expecting to find something after "/foo" (and
    fails to match if nothing is found), while in the second case we can
    match both "/foo" and "/foo/more/to/come". The following are roughly the
    same:

      sub (/foo)   { 'I match /foo' },
      sub (/foo/...) {
        sub (/bar) { 'I match /foo/bar' },
        sub (/*)   { 'I match /foo/{id}' },
      }

    Versus

      sub (/foo...) {
        sub (~)    { 'I match /foo' },
        sub (/bar) { 'I match /foo/bar' },
        sub (/*)   { 'I match /foo/{id}' },
      }

    You may prefer the latter example should you wish to take advantage of
    subdispatchers to scope common activities. For example:

      sub (/user...) {
        my $user_rs = $schema->resultset('User');
        sub (~) { $user_rs },
        sub (/*) { $user_rs->find($_[1]) },
      }

    You should note the special case path match "sub (~)" which is only
    meaningful when it is contained in this type of path match. It matches
    to an empty path.

   Naming your patch matches
    Any */**/*.*/**.* match can be followed with :name to make it into a
    named match, so:

      sub (/*:one/*:two/*:three/*:four) {
        "I match /1/2/3/4 capturing { one => 1, two =>  2, three => 3, four => 4 }"
      }
  
      sub (/**.*:allofit) {
        "I match anything capturing { allofit => \$whole_path }"
      }

    In the specific case of a simple single-* match, the * may be omitted,
    to allow you to write:

      sub (/:one/:two/:three/:four) {
        "I match /1/2/3/4 capturing { one => 1, two =>  2, three => 3, four => 4 }"
      }

   "/foo" and "/foo/" are different specs
    As you may have noticed with the difference between "sub(/foo/...)" and
    "sub(/foo...)", trailing slashes in path specs are significant. This is
    intentional and necessary to retain the ability to use relative links on
    websites. Let's demonstrate on this link:

      <a href="bar">bar</a>

    If the user loads the url "/foo/" and clicks on this link, they will be
    sent to "/foo/bar". However when they are on the url "/foo" and click
    this link, then they will be sent to "/bar".

    This makes it necessary to be explicit about the trailing slash.

   Extension matches
      sub (.html) {

    will match .html from the path (assuming the subroutine itself returns
    something, of course). This is normally used for rendering - e.g.

      sub (.html) {
        response_filter { $self->render_html($_[1]) }
      }

    Additionally,

      sub (.*) {

    will match any extension and supplies the extension as a match argument.

   Query and body parameter matches
    Query and body parameters can be match via

      sub (?<param spec>) { # match URI query
      sub (%<param spec>) { # match body params

    The body spec will match if the request content is either
    application/x-www-form-urlencoded or multipart/form-data - the latter of
    which is required for uploads - see below.

    The param spec is elements of one of the following forms -

      param~        # optional parameter
      param=        # required parameter
      @param~       # optional multiple parameter
      @param=       # required multiple parameter
      :param~       # optional parameter in hashref
      :param=       # required parameter in hashref
      :@param~      # optional multiple in hashref
      :@param=      # required multiple in hashref
      *             # include all other parameters in hashref
      @*            # include all other parameters as multiple in hashref

    separated by the & character. The arguments added to the request are one
    per non-:/* parameter (scalar for normal, arrayref for multiple), plus
    if any :/* specs exist a hashref containing those values.

    Please note that if you specify a multiple type parameter match, you are
    ensured of getting an arrayref for the value, EVEN if the current
    incoming request has only one value. However if a parameter is specified
    as single and multiple values are found, the last one will be used.

    For example to match a page parameter with an optional order_by
    parameter one would write:

      sub (?page=&order_by~) {
        my ($self, $page, $order_by) = @_;
        return unless $page =~ /^\d+$/;
        $page ||= 'id';
        response_filter {
          $_[1]->search_rs({}, $p);
        }
      }

    to implement paging and ordering against a DBIx::Class::ResultSet
    object.

    Another Example: To get all parameters as a hashref of arrayrefs, write:

      sub(?@*) {
        my ($self, $params) = @_;
        ...

    To get two parameters as a hashref, write:

      sub(?:user~&:domain~) {
        my ($self, $params) = @_; # params contains only 'user' and 'domain' keys

    You can also mix these, so:

      sub (?foo=&@bar~&:coffee=&@*) {
         my ($self, $foo, $bar, $params);

    where $bar is an arrayref (possibly an empty one), and $params contains
    arrayref values for all parameters -not- mentioned and a scalar value
    for the 'coffee' parameter.

    Note, in the case where you combine arrayref, single parameter and named
    hashref style, the arrayref and single parameters will appear in @_ in
    the order you defined them in the protoype, but all hashrefs will merge
    into a single $params, as in the example above.

   Upload matches
      sub (*foo=) { # param specifier can be anything valid for query or body

    The upload match system functions exactly like a query/body match,
    except that the values returned (if any) are "Web::Dispatch::Upload"
    objects.

    Note that this match type will succeed in two circumstances where you
    might not expect it to - first, when the field exists but is not an
    upload field and second, when the field exists but the form is not an
    upload form (i.e. content type "application/x-www-form-urlencoded"
    rather than "multipart/form-data"). In either of these cases, what
    you'll get back is a "Web::Dispatch::NotAnUpload" object, which will
    "die" with an error pointing out the problem if you try and use it. To
    be sure you have a real upload object, call

      $upload->is_upload # returns 1 on a valid upload, 0 on a non-upload field

    and to get the reason why such an object is not an upload, call

      $upload->reason # returns a reason or '' on a valid upload.

    Other than these two methods, the upload object provides the same
    interface as Plack::Request::Upload with the addition of a stringify to
    the temporary filename to make copying it somewhere else easier to
    handle.

   Combining matches
    Matches may be combined with the + character - e.g.

      sub (GET + /user/*) {

    to create an AND match. They may also be combined withe the | character
    - e.g.

      sub (GET|POST) {

    to create an OR match. Matches can be nested with () - e.g.

      sub ((GET|POST) + /user/*) {

    and negated with ! - e.g.

      sub (!/user/foo + /user/*) {

    ! binds to the immediate rightmost match specification, so if you want
    to negate a combination you will need to use

      sub ( !(POST|PUT|DELETE) ) {

    and | binds tighter than +, so

      sub ((GET|POST) + /user/*) {

    and

      sub (GET|POST + /user/*) {

    are equivalent, but

      sub ((GET + /admin/...) | (POST + /admin/...)) {

    and

      sub (GET + /admin/... | POST + /admin/...) {

    are not - the latter is equivalent to

      sub (GET + (/admin/...|POST) + /admin/...) {

    which will never match!

   Whitespace
    Note that for legibility you are permitted to use whitespace -

      sub (GET + /user/*) {

    but it will be ignored. This is because the perl parser strips
    whitespace from subroutine prototypes, so this is equivalent to

      sub (GET+/user/*) {

   Accessing parameters via %_
    If your dispatch specification causes your dispatch subroutine to
    receive a hash reference as its first argument, the contained named
    parameters will be accessible via %_.

    This can be used to access your path matches, if they're named:

      sub (GET + /foo/:path_part) {
        [ 200,
          ['Content-type' => 'text/plain'],
          ["We are in $_{path_part}"],
        ];
      }

    Or, if your first argument would be a hash reference containing named
    query parameters:

      sub (GET + /foo + ?:some_param=) {
        [ 200,
          ['Content-type' => 'text/plain'],
          ["We received $_{some_param} as parameter"],
        ];
      }

    Of course this also works when all you are doing is slurping the whole
    set of parameters by their name:

      sub (GET + /foo + ?*) {
        [ 200,
          ['Content-type' => 'text/plain'],
          [exists($_{foo}) ? "Received a foo: $_{foo}" : "No foo!"],
        ],
      }

    Note that only the first hash reference will be avaialble via %_. If you
    receive additional hash references, you will need to access them as
    usual.

   Accessing the PSGI env hash
    In some cases you may wish to get the raw PSGI env hash - to do this,
    you can either use a plain sub -

      sub {
        my ($env) = @_;
        ...
      }

    or use the PSGI_ENV constant exported to retrieve it:

      sub (GET + /foo + ?some_param=) {
        my $param = $_[1];
        my $env = $_[PSGI_ENV];
      }

    but note that if you're trying to add a middleware, you should simply
    use Web::Simple's direct support for doing so.

EXPORTED SUBROUTINES
  response_filter
      response_filter {
        # Hide errors from the user because we hates them, preciousss
        if (ref($_[0]) eq 'ARRAY' && $_[0]->[0] == 500) {
          $_[0] = [ 200, @{$_[0]}[1..$#{$_[0]}] ];
        }
        return $_[0];
      };

    The response_filter subroutine is designed for use inside dispatch
    subroutines.

    It creates and returns a special dispatcher that always matches, and
    calls the block passed to it as a filter on the result of running the
    rest of the current dispatch chain.

    Thus the filter above runs further dispatch as normal, but if the result
    of dispatch is a 500 (Internal Server Error) response, changes this to a
    200 (OK) response without altering the headers or body.

  redispatch_to
      redispatch_to '/other/url';

    The redispatch_to subroutine is designed for use inside dispatch
    subroutines.

    It creates and returns a special dispatcher that always matches, and
    instead of continuing dispatch re-delegates it to the start of the
    dispatch process, but with the path of the request altered to the
    supplied URL.

    Thus if you receive a POST to '/some/url' and return a redispatch to
    '/other/url', the dispatch behaviour will be exactly as if the same POST
    request had been made to '/other/url' instead.


Changes
0.020 - 2012-08-03
  - re-dist for botched release

0.019 - 2012-07-30
  - Add %_ alias if first argument is a hashref
  - Un-mark upload matches as experimental since they work

0.018 - 2012-07-15
  - Allow dots in parameter names
  - Support for basic authentication in test requests

0.017 - 2012-07-05
  - Don't falsely recognize mod_cgid as FCGI

0.016 - 2012-05-11
  - Actually import weaken so the 0.015 fix doesn't implode

0.015 - 2012-05-10
  - Fix a per-application-instance memory leak

0.014 - 2012-04-27
  - Add named path matching
  - Allow headers on CLI calls

0.013 - 2012-04-03
  - Fix a weird interaction between match-spec parsing and module loading
    that was causing brokenness and test failures with perl 5.8.8

0.012 - 2012-01-30
  - Added match predicates match_true and match_false
  - Added '~' to match an empty path
  - Sub-dispatch via '...' is now permissible without a trailing slash
  - Only word char sequences with singular periods are parsed as path parts now
  - Various doc fixes
  - All query body parameters are now decoded from UTF8
  - Allow dots in path matches, so you can have sub (/foo.html)

0.011 - 2011-12-23
  - Add Web::Dispatch::HTTPMethods
  - Fix [ sub {} ] for async code
  - Support match_foo(...) => sub {} as a dispatcher return

0.010 - 2011-10-06
  - Provide $app->run_test_request for use in test scripts
  - Preserve original env in Web::Dispatch so ParamParser can cache in there
    (stops HTTP::Body object getting destroyed early thereby losing uploads)

0.009 - 2011-10-03
  - Complete port from HTTP::Request::AsCGI to Plack::Test
  - Add experimental upload support
  - Update Plack usage to call ::Handler:: classes not ::Server::
  - Assume FastCGI mode if STDIN is a socket (works some places env vars fail)
  - Change CLI mode to print status line and headers to STDERR and content
    to STDOUT so that './myapp /foo.html >foo.html' works sanely
  - Add *.* and **.* dispatch types to keep extension
  - Add Antiquated Perl slides in a POD document.

0.008 - 2011-02-16
  - Once more.

0.007 - 2011-02-16
  - Re-release due to indexing failure

0.006 - 2011-02-15
  - Fix Content-type issue (ie, Content-type: blah works,
    but Content-type: blah; charset: blah shits flames)
  - Make Web::Dispatch return [\$cv] as \$cv to allow subref responses
  - Stop undef errors killing XML::Tags conversion to string
  - Fixup dispatcher logic so dispatchers within a /foo/... work correctly

0.005 - Tue Jan 11 2011 22:09 UTC
  - Redocument to cover changes
  - Factor dispatcher code out into Web::Dispatch
  - Support 'use Web::Simple;' to default to current package

0.004 - Thu Jul 08 2010 22:08 UTC
  - Hide Plack Modules from PAUSE

0.003 - Thu Jul 08 2010 14:48 UTC
  - Experimentally use $_[ENV] for the PSGI env
  - Add CSS::Declare
  - Add more tags to HTML::Tags

0.002 - Tue Dec 01 2009 00:30 UTC
  - fix dispatcher construction to recognise '' as no proto (we expected undef)
  - plackup support, as_psgi_app method
  - Fix SYNOPSIS

0.001 - Tue Nov 24 2009 21:54 UTC

  - Initial release


lib/Web/Simple/Application.pm
package Web::Simple::Application;

use Scalar::Util 'weaken';

use Moo;

has 'config' => (
  is => 'ro',
  default => sub {
    my ($self) = @_;
    +{ $self->default_config }
  },
  trigger => sub {
    my ($self, $value) = @_;
    my %default = $self->default_config;
    my @not = grep !exists $value->{$_}, keys %default;
    @{$value}{@not} = @default{@not};
  }
);

sub default_config { () }

has '_dispatcher' => (is => 'lazy');

sub _build__dispatcher {
  my $self = shift;
  require Web::Dispatch;
  require Web::Simple::DispatchNode;
  my $final = $self->_build_final_dispatcher;

  # We need to weaken both the copy of $self that the
  # app parameter will close over and the copy that'll
  # be passed through as a node argument.
  #
  # To ensure that this doesn't then result in us being
  # DESTROYed unexpectedly early, our to_psgi_app method
  # closes back over $self

  weaken($self);
  my $node_args = { app_object => $self };
  weaken($node_args->{app_object});
  Web::Dispatch->new(
    app => sub { $self->dispatch_request(@_), $final },
    node_class => 'Web::Simple::DispatchNode',
    node_args => $node_args
  );
}

sub _build_final_dispatcher {
  [ 404, [ 'Content-type', 'text/plain' ], [ 'Not found' ] ]
}

sub run_if_script {
  # ->to_psgi_app is true for require() but also works for plackup
  return $_[0]->to_psgi_app if caller(1);
  my $self = ref($_[0]) ? $_[0] : $_[0]->new;
  $self->run(@_);
}

sub _run_cgi {
  my $self = shift;
  require Plack::Handler::CGI;
  Plack::Handler::CGI->new->run($self->to_psgi_app);
}

sub _run_fcgi {
  my $self = shift;
  require Plack::Handler::FCGI;
  Plack::Handler::FCGI->new->run($self->to_psgi_app);
}

sub to_psgi_app {
  my $self = ref($_[0]) ? $_[0] : $_[0]->new;
  my $app = $self->_dispatcher->to_app;

  # Close over $self to keep $self alive even though
  # we weakened the copies the dispatcher has; the
  # if 0 causes the ops to be optimised away to
  # minimise the performance impact and avoid void
  # context warnings while still doing the closing
  # over part. As Mithaldu said: "Gnarly." ...

  return sub { $self if 0; goto &$app; };
}

sub run {
  my $self = shift;
  if (
    $ENV{PHP_FCGI_CHILDREN} || $ENV{FCGI_ROLE} || $ENV{FCGI_SOCKET_PATH}
    || ( -S STDIN && !$ENV{GATEWAY_INTERFACE} )
    # If STDIN is a socket, almost certainly FastCGI, except for mod_cgid
    ) {
    return $self->_run_fcgi;
  } elsif ($ENV{GATEWAY_INTERFACE}) {
    return $self->_run_cgi;
  }
  unless (@ARGV && $ARGV[0] =~ m{(^[A-Z/])|\@}) {
    return $self->_run_cli(@ARGV);
  }

  my @args = @ARGV;

  unshift(@args, 'GET') if $args[0] !~ /^[A-Z]/;

  $self->_run_cli_test_request(@args);
}

sub _test_request_spec_to_http_request {
  my ($self, $method, $path, @rest) = @_;

  # if it's a reference, assume a request object
  return $method if ref($method);

  if ($path =~ s/^(.*?)\@//) {
    my $basic = $1;
    require MIME::Base64;
    unshift @rest, 'Authorization:', 'Basic '.MIME::Base64::encode($basic);
  }

  my $request = HTTP::Request->new($method => $path);

  my @params;

  while (my ($header, $value) = splice(@rest, 0, 2)) {
    unless ($header =~ s/:$//) {
      push @params, $header, $value;
    }
    $header =~ s/_/-/g;
    if ($header eq 'Content') {
      $request->content($value);
    } else {
      $request->headers->push_header($header, $value);
    }
  }

  if (($method eq 'POST' or $method eq 'PUT') and @params) {
    my $content = do {
      require URI;
      my $url = URI->new('http:');
      $url->query_form(@params);
      $url->query;
    };
    $request->header('Content-Type' => 'application/x-www-form-urlencoded');
    $request->header('Content-Length' => length($content));
    $request->content($content);
  }

  return $request;
}

sub run_test_request {
  my ($self, @req) = @_;

  require HTTP::Request;

  require Plack::Test;

  my $request = $self->_test_request_spec_to_http_request(@req);

  Plack::Test::test_psgi(
    $self->to_psgi_app, sub { shift->($request) }
  );
}

sub _run_cli_test_request {
  my ($self, @req) = @_;
  my $response = $self->run_test_request(@req);

  binmode(STDOUT); binmode(STDERR); # for win32

  print STDERR $response->status_line."\n";
  print STDERR $response->headers_as_string("\n")."\n";
  my $content = $response->content;
  $content .= "\n" if length($content) and $content !~ /\n\z/;
  print STDOUT $content if $content;
}

sub _run_cli {
  my $self = shift;
  die $self->_cli_usage;
}

sub _cli_usage {
  "To run this script in CGI test mode, pass a URL path beginning with /:\n".
  "\n".
  "  $0 /some/path\n".
  "  $0 /\n"
}

1;

=head1 NAME

Web::Simple::Application - A base class for your Web-Simple application

=head1 DESCRIPTION

This is a base class for your L<Web::Simple> application.  You probably don't
need to construct this class yourself, since L<Web::Simple> does the 'heavy
lifting' for you in that regards.

=head1 METHODS

This class exposes the following public methods.

=head2 default_config

Merges with the C<config> initializer to provide configuration information for
your application.  For example:

  sub default_config {
    (
      title => 'Bloggery',
      posts_dir => $FindBin::Bin.'/posts',
    );
  }

Now, the C<config> attribute of C<$self>  will be set to a HashRef
containing keys 'title' and 'posts_dir'.

The keys from default_config are merged into any config supplied, so
if you construct your application like:

  MyWebSimpleApp::Web->new(
    config => { title => 'Spoon', environment => 'dev' }
  )

then C<config> will contain:

  {
    title => 'Spoon',
    posts_dir => '/path/to/myapp/posts',
    environment => 'dev'
  }

=head2 run_if_script

The run_if_script method is designed to be used at the end of the script
or .pm file where your application class is defined - for example:

  ## my_web_simple_app.pl
  #!/usr/bin/env perl
  use Web::Simple 'HelloWorld';

  {
    package HelloWorld;

    sub dispatch_request {
      sub (GET) {
        [ 200, [ 'Content-type', 'text/plain' ], [ 'Hello world!' ] ]
      },
      sub () {
        [ 405, [ 'Content-type', 'text/plain' ], [ 'Method not allowed' ] ]
      }
    }
  }

  HelloWorld->run_if_script;

This returns a true value, so your file is now valid as a module - so

  require 'my_web_simple_app.pl';

  my $hw = HelloWorld->new;

will work fine (and you can rename it to lib/HelloWorld.pm later to make it
a real use-able module).

However, it detects if it's being run as a script (via testing $0) and if
so attempts to do the right thing.

If run under a CGI environment, your application will execute as a CGI.

If run under a FastCGI environment, your application will execute as a
FastCGI process (this works both for dynamic shared-hosting-style FastCGI
and for apache FastCgiServer style setups).

If run from the commandline with a URL path, it runs a GET request against
that path -

  $ perl -Ilib examples/hello-world/hello-world.cgi /
  200 OK
  Content-Type: text/plain
  
  Hello world!

You can also provide a method name -

  $ perl -Ilib examples/hello-world/hello-world.cgi POST /
  405 Method Not Allowed
  Content-Type: text/plain
  
  Method not allowed

For a POST or PUT request, pairs on the command line will be treated
as form variables. For any request, pairs on the command line ending in :
are treated as headers, and 'Content:' will set the request body -

  $ ./myapp POST / Accept: text/html form_field_name form_field_value
  
  $ ./myapp POST / Content-Type: text/json Content: '{ "json": "here" }'

The body of the response is sent to STDOUT and the headers to STDERR, so

  $ ./myapp GET / >index.html

will generally do the right thing.

To send basic authentication credentials, use user:pass@ syntax -

  $ ./myapp GET bob:secret@/protected/path

Additionally, you can treat the file as though it were a standard PSGI
application file (*.psgi).  For example you can start up up with C<plackup>

  plackup my_web_simple_app.pl

or C<starman>

  starman my_web_simple_app.pl

=head2 to_psgi_app

This method is called by L</run_if_script> to create the L<PSGI> app coderef
for use via L<Plack> and L<plackup>. If you want to globally add middleware,
you can override this method:

  use Web::Simple 'HelloWorld';
  use Plack::Builder;
 
  {
    package HelloWorld;

  
    around 'to_psgi_app', sub {
      my ($orig, $self) = (shift, shift);
      my $app = $self->$orig(@_); 
      builder {
        enable ...; ## whatever middleware you want
        $app;
      };
    };
  }

This method can also be used to mount a Web::Simple application within
a separate C<*.psgi> file -

  use strictures 1;
  use Plack::Builder;
  use WSApp;
  use AnotherWSApp;

  builder {
    mount '/' => WSApp->to_psgi_app;
    mount '/another' => AnotherWSApp->to_psgi_app;
  };

This method can be called as a class method, in which case it implicitly
calls ->new, or as an object method ... in which case it doesn't.

=head2 run

Used for running your application under stand-alone CGI and FCGI modes.

I should document this more extensively but run_if_script will call it when
you need it, so don't worry about it too much.

=head2 run_test_request

  my $res = $app->run_test_request(GET => '/' => %headers);

  my $res = $app->run_test_request(POST => '/' => %headers_or_form);

  my $res = $app->run_test_request($http_request);

Accepts either an L<HTTP::Request> object or ($method, $path) and runs that
request against the application, returning an L<HTTP::Response> object.

If the HTTP method is POST or PUT, then a series of pairs can be passed after
this to create a form style message body. If you need to test an upload, then
create an L<HTTP::Request> object by hand or use the C<POST> subroutine
provided by L<HTTP::Request::Common>.

If you prefix the URL with 'user:pass@' this will be converted into
an Authorization header for HTTP basic auth:

  my $res = $app->run_test_request(
              GET => 'bob:secret@/protected/resource'
            );

If pairs are passed where the key ends in :, it is instead treated as a
headers, so:

  my $res = $app->run_test_request(
              POST => '/',
             'Accept:' => 'text/html',
              some_form_key => 'value'
            );

will do what you expect. You can also pass a special key of Content: to
set the request body:

  my $res = $app->run_test_request(
              POST => '/',
              'Content-Type:' => 'text/json',
              'Content:' => '{ "json": "here" }',
            );

=head1 AUTHORS

See L<Web::Simple> for authors.

=head1 COPYRIGHT AND LICENSE

See L<Web::Simple> for the copyright and license.

=cut
